DataStructures/C/LinkedList-in-C/generic-linked-list/ll.c

```diff
@@ -0,0 +1,607 @@
+        prev = curr;
+        curr = curr->next;
+        curr_index++;
+    }
+
+    if (!curr) return head;
+
+    LinkedListNode_t* range_start = prev;
+
+    while (curr && curr_index <= end) {
+        LinkedListNode_t* temp = curr;
+        curr = curr->next;
+
+        deallocator(temp->data);
+        free(temp);
+
+        curr_index++;
+    }
+
+    if (range_start) {
+        range_start->next = curr;
+    } else {
+        head = curr;
+    }
+
+    return head;
+}
+
+LinkedListNode_t* ll_remove_all(LinkedListNode_t* head, void* target, int (*cmp)(void*, void*), void (*deallocator)(void*)) {
+    if (!head || !cmp || !deallocator) return NULL;  // return `NULL` if `head` pointer, `cmp` function or the
+                                                     //`deallocator` function is not provided.
+
+    LinkedListNode_t* prev = NULL;
+    LinkedListNode_t* curr = head;
+
+    while (curr) {
+        if (cmp(curr->data, target) == 0) {  // checks whether the `curr.data` element is equal to `target` or not.
+            if (prev) {
+                prev->next = curr->next;
+            } else {
+                head = curr->next;
+            }
+
+            deallocator(curr->data);
+            free(curr);
+        }
+        prev = curr;
+        curr = curr->next;
+    }
+
+    return head;
+}
+
+LinkedListNode_t* ll_sort(LinkedListNode_t* head, int (*cmp)(void*, void*)) {
+    if (!cmp) return NULL;  // return `NULL` incase if `cmp` is not provided.
+
+    return __ll_merge_sort(head, cmp);
+}
+
+LinkedListNode_t* ll_reverse(LinkedListNode_t* head) {
+    LinkedListNode_t* prev = NULL;
+    LinkedListNode_t* curr = head;
+    LinkedListNode_t* next = NULL;
+
+    while (curr) {
+        next = curr->next;
+        curr->next = prev;
+
+        prev = curr;
+        curr = next;
+    }
+
+    return prev;
+}
+
+LinkedListNode_t* ll_map(LinkedListNode_t* head, void* (*fx)(void*)) {
+    if (!head || !fx) return NULL;  // return `NULL` incase if `head` or `fx` are not provided or are `NULL`.
+
+    LinkedListNode_t* new_head = ll_new_node(fx(head->data));
+    LinkedListNode_t* new_tail = new_head;
+    LinkedListNode_t* node = head->next;
+
+    while (node) {
+        new_tail->next = ll_new_node(fx(node->data));
+        new_tail = new_tail->next;
+        node = node->next;
+    }
+
+    return new_head;
+}
+
+LinkedListNode_t* ll_filter(LinkedListNode_t* head, bool (*predicate)(void*), void* (*copy_element)(void*)) {
+    if (!head || !predicate) return NULL;  // return `NULL` incase if `head` or `predicate` are not provided or are
+                                           // `NULL`.
+
+    LinkedListNode_t* new_head = NULL;
+    LinkedListNode_t* new_tail = NULL;
+
+    for (LinkedListNode_t* node = head; node; node = node->next) {
+        if (predicate(node->data)) {
+            void* new_data = copy_element ? copy_element(node->data) : node->data;  // if `copy_element` is `NULL` then
```

Comment:
```
if (!copy_element) {
    return NULL; // Explicitly require a copy function to prevent accidental shallow copies.
}
void* new_data = copy_element(node->data);
```